File: 0009-group-anagrams/solution.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

func groupAnagrams(strs []string) [][]string {
	dict := make(map[string][]string)

	for _, str := range strs {
		chars := strings.Split(str, "")
		sort.Strings(chars)

		key := strings.Join(chars, "")

		dict[key] = append(dict[key], str)
	}

	result := make([][]string, 0, len(dict))

	for _, group := range dict {
		result = append(result, group)
	}

	return result
}

func groupAnagramsUsingHash(strs []string) [][]string {
	dict := make(map[[26]int][]string)

	for _, str := range strs {

		var countArray [26]int

		fmt.Println("Count Array Before for str:", countArray)

		for _, char := range str {
			index := char - 'a'
			countArray[index]++
		}

		fmt.Println("Count Array After for str:", countArray)

		dict[countArray] = append(dict[countArray], str)
	}

	result := make([][]string, 0, len(dict))

	for _, group := range dict {
		result = append(result, group)
	}

	return result
}

File: 0009-group-anagrams/solution.swift
func groupAnagrams(_ strs: [String]) -> [[String]] {
    var dict: [String: [String]] = [:]
    
    for str in strs {
        let key = String(str.sorted())
        
        dict[key, default: []].append(str)
    }
    
    return dict.values.map { $0 }
}
//n = number of strings
//k = maximum length of a string
//Time:  O(n × k log k)
//Space: O(n × k)

func groupAnagramsUsingHash(_ strs: [String]) -> [[String]] {
    var dict: [[Int]: [String]] = [:]
    
    for str in strs {
        var countArray = Array(repeating: 0, count: 26)
        print("Count Array Before or str: \(countArray)")
        
        for char in str {
            let index = Int(char.asciiValue! - Character("a").asciiValue!)
            countArray[index] += 1
        }
        print("Count Array After or str: \(countArray)")
        dict[countArray, default: []].append(str)
    }
    
    return Array(dict.values)
}

//Time:  O(n × k)
//Space: O(n × k)
